3.0 - Improved belongsToMany select field support

ndm2 opened the pull request:
I'm not too sure about the magic options, as 3.x was supposed to have less magic, however it's just a tiny bit of magic, and since similar is supported for belongsTo associations, this might be within reason.

Commit:
Add magic options support for belongsToMany selects

This already works for belongsTo associations, making it work for
belongsToMany might come in handy too.

Using `substr()` as in other places, it's way faster than a regex,
and there is no complexity involved that would make a regex
necessary.

Commit:
Add support for recognizing errors on belongsToMany selects

While it is possible to validate fields for belongsToMany selects,
this must be done using the plain association property, and so the
form helper will not able to pick up possible errors, as it will
check for the fieldname with `._ids` appended.

Commit:
Don't do unnecessary checks

markstory commented on src/View/Helper/FormHelper.php:
```diff
@@ -1167,9 +1170,14 @@ protected function _optionsOptions($fieldName, $options)
+            $fieldName = substr($fieldName, 0, -5);
+        } elseif (substr($fieldName, -3) === '_id') {
+            $fieldName = substr($fieldName, 0, -3);
+        }
```
Why is this necessary? Line 1170 currently takes the first segment which would be `tags` in the case of `tags._ids`.

ndm2 replied:
Actually it takes the last segment.

markstory commented on src/View/Helper/FormHelper.php:
```diff
@@ -1167,9 +1170,14 @@ protected function _optionsOptions($fieldName, $options)
         $fieldName = array_slice(explode('.', $fieldName), -1)[0];
         $varName = Inflector::variable(
-            Inflector::pluralize(preg_replace('/_id$/', '', $fieldName))
+            Inflector::pluralize($fieldName)
```
This isn't going to work for many scenarios as there are a huge pile of words that cannot be pluralized twice. I think we're just getting 'lucky' with tags.  I'd rather have no magic vs. inconsistent magic.

ndm2 replied:
Ah, right, I vaguely remember that this caused problems here and there before :)

ndm2 replied:
Manually changed association properties and foreign keys aside - As per conventions, it should only be necessary to pluralize belongsTo/hasOne fields, shouldn't it? If that is the case, could pluralizing based on the postfix be a viable solution (given that it's worth to keep the magic anyways)?

markstory replied:
Pluralizing based on `_id` but not `._ids` seems within reason to me.

ndm2 replied:
Meh, looking at the tests, namely [`testInputOverridingMagicSelectType`](https://github.com/ndm2/cakephp/blob/3.0-improved-belongs-to-many-select-field-support/tests/TestCase/View/Helper/FormHelperTest.php#L2934) and [`testInputMagicSelectForTypeNumber`](https://github.com/ndm2/cakephp/blob/3.0-improved-belongs-to-many-select-field-support/tests/TestCase/View/Helper/FormHelperTest.php#L3033), it seems that the general pluralization to magically pick up any variable based on this convention is still wanted?

markstory replied:
Wanted-ish. I think I tried to remove it at some point and people were not happy about that.

ndm2 replied:
I see... guess I'll have to go with only excluding `._ids` postfixed fieldnames then.

Commit:
Exclude `._ids` postfixed fieldnames from pluralization

As per convention, fieldnames for belongsToMany associations should
be expected to be already in plural form.

markstory commented on src/View/Helper/FormHelper.php:
```diff
@@ -674,6 +674,9 @@ public function isFieldError($field)
     {
+        if (substr($field, -5) === '._ids') {
+            $field = substr($field, 0, -5);
+        }
```
Will the array property ever have a validation error attached to it?

ndm2 replied:
It might have, say for example you wanted to validate that a maximum/minimum of x tags is being selected.

markstory commented on tests/TestCase/View/Helper/FormHelperTest.php:
```diff
@@ -4126,6 +4126,79 @@ public function testHabtmSelectBox()
+            1 => 'blue',
+            50 => 'green'
+        ];
+        $this->View->viewVars['tags'] = $tags;
```
It might be better to use something like `people` here as that is a name that would break if passed through pluralize().

ndm2 replied:
Indeed

Commit:
Test with a variable name that would break when pluralized

Commit:
Why not both

markstory commented:
Looks good to me.
